ouroboros-consensus/docs/VersioningSchemeDecision.md

```diff
@@ -0,0 +1,271 @@
+Prelude Data.Version> makeVersion [1,2] < makeVersion [1,2,0]
+True
+```
+
+Each `main` version has only two dimensions: `A.B`.
+Each release version has at least three `A.B.C`, where `C` can be `0`.
+
+PRs do not alter the `main` version.
+
+To cut a release from a commit COMMIT on `main` that declares version `A.B`, announce a new non-`main` commit that extends COMMIT merely to declare version `A.B.0`.
+Also immediately update the `main` version to `next(A.B)`.
+
+Pros:
+
+- It enforces all desired invariants.
+
+- It's a very mechanical state machine, easy to execute and also easy to immediately recognize which state it's in.
+
+Cons:
+
+- The multi-sorted state transition system prevents any explanation from being comparatively small.
+
+- This scheme is certainly not already well-established!
+
+- It incurs spurious increments of `A.B` when the only differences between two releases were patch PRs.
+
+## Proposal Dimension124
+
+Proposal Dimension above has the downside that it incurs spurious increments of `A.B` when the only differences between two releases were patch PRs.
+The following enrichment adds the minimal amount of additional complexity to avoid that without losing any invariants.
+
+Each `main` version is `A`, `A.B` where `B>0`, or `A.B.C.2718`; it is never three-dimensional.
+Each release version is the usual `A.B.C`.
+`2718` is just a recognizable magic number; it's so large it's unlikely to come up in actual versioning and it's the first digits of _e_, which is related to _growth_, which is what the `main` branch is for.
+
+PRs alter the `main` version as indicated in the following diagram.
+
+```
+  A          --------[A-level PR]---------> A
+  A          --------[B-level PR]---------> A
+  A          --------[C-level PR]---------> A
+
+  A.B        --------[A-level PR]---------> A+1
+  A.B        --------[B-level PR]---------> A.B
+  A.B        --------[C-level PR]---------> A.B
+
+  A.B.C.2718 --------[A-level PR]---------> A+1
+  A.B.C.2718 --------[B-level PR]---------> A.(B+1)
+  A.B.C.2718 --------[C-level PR]---------> A.B.C.2718
+```
+
+To cut a release from a commit COMMIT on `main` that declares version `A`, announce a new non-`main` commit that extends COMMIT merely to declare version `A.0.0`.
+Also immediately update the `main` version to `A.0.0.2718`.
+
+To cut a release from a commit COMMIT on `main` that declares version `A.B`, announce a new non-`main` commit that extends COMMIT merely to declare version `A.B.0`.
+Also immediately update the `main` version to `A.B.0.2718`.
+
+To cut a release from a commit COMMIT on `main` that declares version `A.B.C.2718`, announce a new non-`main` commit that extends COMMIT merely to declare version `A.B.(C+1)`.
+Also immediately update the `main` version to `A.B.(C+1).2718`.
+
+The above is summarized in the following diagram, which relies on this legend depicting the semantics of the one kind of node (state) and the semantics of the two kinds of edges (transitions).
```

Comment:
I'm quite proud of this Mermaid diagram. I think it does a really good job of explaining the proposal. The only two improvements I can imagine:

- better node positioning (I don't know how with Mermaid)
- better classifications of the "merge a PR" transitions; "bugfix", "backwards-compat" and "backwards-incompat" are familiar and suggestive but perhaps overly-precise?

Reply:
The diagram definitely helped me to understand the approach being proposed.